### PR TITLE
Implement `textDocument/selectionRange`

### LINKED ISSUE #1021: Support for 'textDocument/selectionRange'
Link to spec: https://microsoft.github.io/language-server-protocol/specification#textDocument_selectionRange

Example of what it enables: https://rust-analyzer.github.io/manual.html#expand-and-shrink-selection

### PR DESCRIPTION
**Pull request by datho7561**
Requires the client to be updated to support LSP 3.15

Closes #1021

Signed-off-by: David Thompson <davthomp@redhat.com>

### CODE REVIEW

**Comment by datho7561** [comment]
Here is a example of what you can do now in this pr:

selection-range-demo

I submitted this as a draft, since its mostly working, but I ran into some corner cases that might take a while to figure out.

File: org.eclipse.lemminx/src/test/java/org/eclipse/lemminx/services/XMLSelectionRangesTest.java [hunk 1]
```diff
@@ -0,0 +1,506 @@
...
+				sr(r(0, 15, 1, 0), r(0, 0, 1, 2), r(0, 0, 5, 0)));
+	}
+
+	//#endregion
+
+	//#region test utils
+
+	private static final CancelChecker NULL_CHECKER = () -> {};
+
+	private void testSelectionRange(String xml, SelectionRange... selectionRanges) {
```
Comment (angelozerr) [inline org.eclipse.lemminx/src/test/java/org/eclipse/lemminx/services/XMLSelectionRangesTest.java:459]: Please move testSelectionRange in XMLAssert to be consistent with other features.

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/services/XMLSelectionRanges.java [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+		leafSelectionRange.setRange(new Range());
+		SelectionRange selectionRange = leafSelectionRange;
+
+		do {
+			selectionRange = handleNodeSelectionRange(selectionRange, node, offset);
+			if (node.isAttribute() && ((DOMAttr)node).getOwnerElement() != null) {
+				node = ((DOMAttr)node).getOwnerElement();
+			}
+			DOMNode prolog = node.getOwnerDocument().getProlog();
+			if (prolog != null && node != node.getOwnerDocument() && node != prolog && prolog.getStart() <= offset && offset <= prolog.getEnd()) {
```
Comment (datho7561) [inline org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/services/XMLSelectionRanges.java:94]: I'll clean up this line so its easier to read

**Comment by datho7561** [comment]
Please do not merge; I figured out the bug that is causing the tests I disabled, and am going to fix it

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/services/XMLSelectionRanges.java [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+		selectionRange = handleGenericNodeSelectionRange(selectionRange, attr, offset);
+		DOMElement ownerElement = attr.getOwnerElement();
+		if (ownerElement != null) {
+			selectionRange = handleGenericNodeSelectionRange(selectionRange, ownerElement, offset);
+		}
+		return selectionRange;
+	}
+
+	private SelectionRange handleElementSelectionRange(SelectionRange selectionRange, DOMElement element, int offset) {
+		XMLPositionUtility.selectStartTagName(element);
```
Comment (rgrunber) [inline org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/services/XMLSelectionRanges.java:145]: Does this have a side effect ? Seems to only be useful for its returned value.

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/services/XMLSelectionRanges.java [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+		} else if (contentStart != DOMNode.NULL_VALUE && offset < contentStart) {
+			SelectionRange parentSelectionRange = new SelectionRange();
+			parentSelectionRange.setRange(XMLPositionUtility.selectStartTagName(element));
+			selectionRange.setParent(parentSelectionRange);
+			selectionRange = parentSelectionRange;
+		} else if (offset <= element.getEndTagCloseOffset()) {
+			SelectionRange parentSelectionRange = new SelectionRange();
+			parentSelectionRange.setRange(XMLPositionUtility.selectEndTagName(element));
+			selectionRange.setParent(parentSelectionRange);
+			selectionRange = parentSelectionRange;
```
Comment (rgrunber) [inline org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/services/XMLSelectionRanges.java:166]: Would it make more sense to move :

`SelectionRange parentSelectionRange = new SelectionRange();`

above the start of the if statement and :

`selectionRange.setParent(parentSelectionRange);`
`selectionRange = parentSelectionRange;`

below the if block ?

Those statements seem to stay the same.

This would leave just `parentSelectionRange.setRange(...)` within the conditionals.

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/services/XMLSelectionRanges.java [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+		return handleGenericNodeSelectionRange(selectionRange, element, offset);
+	}
+
+	private SelectionRange handleDoctypeSelectionRange(SelectionRange selectionRange, DOMDocumentType doctype,
+			int offset) {
+		DOMRange internalSubset = doctype.getInternalSubsetNode();
+		if (internalSubset != null) {
+			int subsetStart = internalSubset.getStart();
+			int subsetEnd = internalSubset.getEnd();
+			// (subsetStart != -1 && subsetStart < offset && offset < subset)
```
Comment (rgrunber) [inline org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/services/XMLSelectionRanges.java:177]: Stil needed ?

**Review by rgrunber — COMMENTED** [review]
Works for me! Just a few small things to look at.

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/services/XMLSelectionRanges.java (hunk 3 above)
Comment (datho7561) [reply org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/services/XMLSelectionRanges.java:145]: No, this won't do anything. I'll remove it.

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/services/XMLSelectionRanges.java (hunk 5 above)
Comment (datho7561) [reply org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/services/XMLSelectionRanges.java:177]: This was just an explanation of why I didn't check `subsetEnd != -1`. I'll clean up the wording.

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/services/XMLSelectionRanges.java (hunk 4 above)
Comment (datho7561) [reply org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/services/XMLSelectionRanges.java]: There are some cases where I want to avoid adding a new parent selection range, for instance when the cursor is here:
```xml
...
  |<element>content</element>
...
```
I just want the range that covers the entire element, so I don't add a step before selecting the entire element.

I moved the `parentSelectionRange` creation to the beginning, since it can be GC'd and it makes the code easier to read

File: org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/dom/DOMAttr.java [hunk 6]
```diff
@@ -62,6 +62,11 @@ public DOMAttr getOwnerAttr() {
 			return DOMAttr.this;
 		}
 
+		@Override
+		public DOMNode getParentNode() {
```
Comment (angelozerr) [inline org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/dom/DOMAttr.java:66]: Why you did that? It doesn't follow the w3c spec https://www.w3.org/2003/01/dom2-javadoc/org/w3c/dom/Attr.html

`Thus, the Node attributes parentNode, previousSibling, and nextSibling have a null value for Attr objects.`

Comment (datho7561) [reply org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/dom/DOMAttr.java:66]: This Node is not `Attr` though, this is a child of `Attr`. If the children of `Attr` are also not allowed to have a parent, then I will look into a different implementation of this.

Comment (angelozerr) [reply org.eclipse.lemminx/src/main/java/org/eclipse/lemminx/dom/DOMAttr.java:66]: Ooops forget my comment, sorry.

**Comment by angelozerr** [comment]
@datho7561 I'm OK to merge the PR. Great job!
